drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1279 @@
+		for (k = 0; k < buf->width; ++k) {
+			off = buf->stride * j + k * 4;
+			*(uint32_t*)&buf->map[off] =
+				     (out->r << 16) | (out->g << 8) | out->b;
+		}
+	}
+
+}
+
+/*
+ * modeset_draw_out() prepares the framebuffer with the drawing and then it asks
+ * for the driver to perform an atomic commit. This will lead to a page-flip and
+ * the content of the framebuffer will be displayed. In this simple example
+ * we're only using the primary plane, but we could also be updating other
+ * planes in the same atomic commit.
+ *
+ * Just like in modeset_perform_modeset(), we first setup everything with
+ * modeset_atomic_prepare_commit() and then actually perform the atomic commit.
+ * But there are some important differences:
+ *
+ * 1. Here we just want to perform a commit that changes the state of a specific
+ *    output, and in modeset_perform_modeset() we did an atomic commit that was
+ *    supposed to setup all the outputs at once. So there's no need to prepare
+ *    every output before performing the atomic commit. But let's suppose you
+ *    prepare every output and then perform the commit. It should schedule a
+ *    page-flip for all of them, but modeset_draw_out() was called because the
+ *    page-flip for a specific output has finished. The others may not be
+ *    prepared for a page-flip yet (e.g. in the middle of a scanout), so these
+ *    page-flips will fail.
+ *
+ * 2. Here we have already painted the framebuffer and also we don't use the
+ *    flag DRM_MODE_ALLOW_MODESET anymore, since the modeset already happened.
+ *    We could continue to use this flag, as it makes no difference if
+ *    modeset_perform_modeset() is correct and there's no bug in the kernel.
+ *    The flag only allows (it doesn't force) the driver to perform a modeset,
+ *    but we have already performed it in modeset_perform_modeset() and now we
+ *    just want page-flips to occur. If we still need to perform modesets it
+ *    means that we have a bug somewhere, and it may be better to fail than to
+ *    glitch (a modeset can cause unecessary latency and also blank the screen).
+ */
+
+static void modeset_draw_out(int fd, struct modeset_output *out)
+{
+	drmModeAtomicReq *req;
+	int ret, flags;
+
+	/* draw on framebuffer of the output */
+	modeset_paint_framebuffer(out);
+
+	/* prepare output for atomic commit */
+	req = drmModeAtomicAlloc();
+	ret = modeset_atomic_prepare_commit(fd, out, req);
+	if (ret < 0) {
+		fprintf(stderr, "prepare atomic commit failed, %d\n", errno);
+		return;
+	}
+
+	/* We've just draw on the framebuffer, prepared the commit and now it's
+	 * time to perform a page-flip to display its content.
+	 *
+	 * DRM_MODE_PAGE_FLIP_EVENT signalizes that we want to receive a page-flip
+	 * event in the DRM-fd when the page-flip happens. This flag is also used in
+	 * the non-atomic examples, so you're probably familiar with it.
+	 *
+	 * DRM_MODE_ATOMIC_NONBLOCK makes the page-flip asynchronous. We don't want
```

Comment:
"Asynchronous" and "non-blocking" actually have very specific meanings in KMS.

Asynchronous: do not wait for vblank but flip right now. This can cause tearing.

Non-blocking: return from the ioctl before the operation is complete. Implies some other way of knowing when the operation actually does complete (e.g. pageflip event).

Reply:
Thanks for the explanation. Fixed.